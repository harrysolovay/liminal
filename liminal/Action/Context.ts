import type { Action } from "./Action.js"
import type { ActorLike } from "../common/ActorLike.js"
import { ActionBase } from "./ActionBase.js"
import type { EventBase } from "./event_common.js"
import type { ActionEvent } from "./ActionEvent.js"
import type { Spec } from "../Spec.js"

export function Context<K extends string, Y extends Action, R = string>(
  key: K,
  system: string,
  implementation: ActorLike<Y, R>,
): Generator<
  Context<{
    Model: never
    Event: never
  }>,
  Awaited<R>
>
export function Context<K extends string, Y extends Action, R = string>(
  key: K,
  implementation: ActorLike<Y, R>,
): Generator<
  Context<{
    Model: never
    Event: never
  }>,
  Awaited<R>
>
export function* Context<Y extends Action, R = string>(
  key: string,
  a0: string | ActorLike,
  a1?: ActorLike,
): Generator<Context, Awaited<R>> {
  return yield ActionBase("Context", {
    key,
    system: typeof a0 === "string" ? a0 : undefined,
    implementation: typeof a0 === "string" ? a1 : a0,
  })
}

export interface Context<S extends Spec = Spec> extends ActionBase<"Context", S> {
  key: string
  system: string | undefined
  implementation?: ActorLike
}

export interface ContextEvent<K extends string = string, E extends ActionEvent = ActionEvent>
  extends EventBase<"Context"> {
  key: K
  inner: E
}
