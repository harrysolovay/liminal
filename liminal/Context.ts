import { AsyncLocalStorage } from "node:async_hooks"

const storage = new AsyncLocalStorage<Context>()

export class Context extends Map<ContextPart, unknown> {
  static get(): Context | undefined {
    return storage.getStore()
  }

  run<R>(f: () => R): R {
    return storage.run(this, f)
  }

  fork(): Context {
    const context = new Context()
    for (const [handle, value] of this.entries()) {
      if (!context.has(handle)) {
        context.set(handle, handle.fork(value))
      }
    }
    return context
  }
}

export interface ContextPart<V = any> {
  fork(parent?: V): V
  get(): V | undefined
  debug?: string
}

export function ContextPart<V>(fork: (parent?: V) => V, debug?: string): ContextPart<V> {
  const self: ContextPart<V> = {
    fork,
    get() {
      return Context.get()?.get(self) as never
    },
    ...debug && { debug },
  }
  return self
}
