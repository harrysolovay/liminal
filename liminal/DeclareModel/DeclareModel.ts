import { ActionBase } from "../Action/ActionBase.js"
import type { Spec } from "../Spec.js"
import type { ModelDeclaredEvent } from "./DeclareModelEvent.js"
import type { ModelPurpose } from "./ModelPurpose.js"

export class DeclareModel<S extends Spec = Spec> implements ActionBase<"declare_model", S> {
  static *language<K extends keyof any>(key: K): Generator<
    DeclareModel<{
      LanguageModel: K
      EmbeddingModel: never
      Event: ModelDeclaredEvent<K, "language">
    }>,
    void
  > {
    return yield ActionBase("declare_model", {
      key,
      purpose: "language",
    })
  }

  static *embedding<K extends keyof any>(key: K): Generator<
    DeclareModel<{
      LanguageModel: never
      EmbeddingModel: K
      Event: ModelDeclaredEvent<K, "embedding">
    }>,
    void
  > {
    return yield ActionBase("declare_model", {
      key,
      purpose: "embedding",
    })
  }

  declare "": S
  declare action: "declare_model"
  declare key: keyof any
  declare purpose: ModelPurpose
}
