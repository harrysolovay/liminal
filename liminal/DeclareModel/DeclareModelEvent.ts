import type { ActionEventBase } from "../Action/ActionEventBase.js"
import type { ModelPurpose } from "./ModelPurpose.js"

export interface ModelDeclaredEvent<K extends keyof any = keyof any, P extends ModelPurpose = ModelPurpose>
  extends ActionEventBase<"model_declared">
{
  key: K
  purpose: P
}
