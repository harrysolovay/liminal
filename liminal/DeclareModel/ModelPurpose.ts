export type ModelPurpose = "language" | "embedding"
