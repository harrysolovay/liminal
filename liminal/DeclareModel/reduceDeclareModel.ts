import type { ActionReducer } from "../Action/ActionReducer.ts"
import type { DeclareModel } from "./DeclareModel.ts"

export const reduceDeclareModel: ActionReducer<DeclareModel> = (scope, action) => {
  scope.events.emit({
    type: "model_declared",
    key: action.key,
    purpose: action.purpose,
  })
  return scope.spread({
    next: undefined,
    model: {
      ...scope.model,
      ...(action.purpose === "language"
        ? { language: action.key }
        : { embedding: action.key }),
    },
  })
}
