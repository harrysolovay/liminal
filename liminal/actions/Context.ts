import type { ActionEvent } from "../ActionEvent.ts"
import type { ActionLike } from "../Actor/ActionLike.ts"
import type { ActorLike } from "../Actor/ActorLike.ts"
import { reduce } from "../Actor/reduce.ts"
import { Scope } from "../Scope.ts"
import type { ExtractSpec, Spec } from "../Spec.ts"
import { unwrapDeferred } from "../util/unwrapDeferred.ts"
import { ActionBase, type ActionEventBase, type EnteredEvent, type ExitedEvent } from "./actions_base.ts"

export interface Context<S extends Spec = Spec> extends ActionBase<"context", S> {
  key: keyof any
  implementation: ActorLike
}

export function* context<K extends keyof any, Y extends ActionLike, S extends ExtractSpec<Y>, T = string>(
  key: K,
  implementation: ActorLike<Y, T>,
): Generator<
  Context<{
    Entry: S["Entry"]
    Event: ContextEvent<K, EnteredEvent | S["Event"] | ExitedEvent<T>>
  }>,
  T
> {
  return yield ActionBase("context", {
    key,
    implementation,
    async reduce(scope) {
      const actor = unwrapDeferred(implementation)
      const events = scope.events.child((event) => ({
        type: "context",
        context: key,
        event,
      }))
      events.emit({
        type: "entered",
      })
      const contextScope = await reduce(
        actor,
        new Scope(
          scope.args,
          key,
          events,
          scope.runInfer,
          scope.runEmbed,
        ),
      )
      events.emit({
        type: "exited",
        result: contextScope.result,
      })
      return scope.spread({
        next: contextScope.result,
        children: [...scope.children, {
          type: "context",
          scope: contextScope,
        }],
      })
    },
  })
}

export interface ContextEvent<K extends keyof any = keyof any, E extends ActionEvent = any>
  extends ActionEventBase<"context">
{
  context: K
  event: E
}
