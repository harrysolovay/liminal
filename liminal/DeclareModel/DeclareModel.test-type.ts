import { Context } from "../Context/Context.ts"
import type { ContextEnteredEvent, ContextExitedEvent, ContextInnerEvent } from "../Context/ContextEvent.ts"
import { ActorAssertions } from "../testing/ActorAssertions.ts"
import { DeclareModel } from "./DeclareModel.ts"
import type { ModelDeclaredEvent } from "./DeclareModelEvent.ts"

const languageModel = DeclareModel.language("A")
ActorAssertions(languageModel).assertSpec<{
  LanguageModel: "A"
  EmbeddingModel: never
  Event: ModelDeclaredEvent<"A", "language">
}>()

const embeddingModel = DeclareModel.embedding("B")
ActorAssertions(embeddingModel).assertSpec<{
  LanguageModel: never
  EmbeddingModel: "B"
  Event: ModelDeclaredEvent<"B", "embedding">
}>()

function* both() {
  yield* languageModel
  yield* embeddingModel
}

ActorAssertions(both).assertSpec<{
  LanguageModel: "A"
  EmbeddingModel: "B"
  Event: ModelDeclaredEvent<"A", "language"> | ModelDeclaredEvent<"B", "embedding">
}>()

function* parent() {
  yield* Context("Context", function*() {
    yield* both()
  })
  yield* DeclareModel.language("C")
  yield* DeclareModel.embedding("D")
}

ActorAssertions(parent).assertSpec<{
  LanguageModel: "A" | "C"
  EmbeddingModel: "B" | "D"
  Event:
    | ModelDeclaredEvent<"C", "language">
    | ModelDeclaredEvent<"D", "embedding">
    | ContextEnteredEvent<"Context">
    | ContextInnerEvent<"Context", ModelDeclaredEvent<"A", "language"> | ModelDeclaredEvent<"B", "embedding">>
    | ContextExitedEvent<"Context", void>
}>()
