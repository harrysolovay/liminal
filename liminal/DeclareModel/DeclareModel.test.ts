import { describe, expect, it } from "bun:test"
import { Context } from "../Context/Context.ts"
import { Exec } from "../Exec/Exec.ts"
import { TestEmbeddingModel } from "../testing/TestEmbeddingModel.ts"
import { TestLanguageModel } from "../testing/TestLanguageModel.ts"
import { DeclareModel } from "./DeclareModel.ts"

describe("Model", () => {
  it("generates the expected event sequence", async () => {
    const scope = await Exec(function*() {
      yield* DeclareModel.language("default")
      yield* DeclareModel.language("secondary")
      yield* Context("child", function*() {
        yield* DeclareModel.language("child_a")
        yield* DeclareModel.embedding("child_b")
      })
      yield* DeclareModel.embedding("tertiary")
    })
      .models({
        default: TestLanguageModel(),
        secondary: TestLanguageModel(),
        child_a: TestLanguageModel(),
        child_b: TestEmbeddingModel(),
        tertiary: TestEmbeddingModel(),
      })
      .exec()
    expect(JSON.stringify(scope, null, 2)).toMatchSnapshot()
  })
})
