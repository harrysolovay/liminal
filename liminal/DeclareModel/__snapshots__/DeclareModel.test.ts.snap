// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Model generates the expected event sequence 1`] = `
"{
  "events": [
    {
      "type": "exec_entered"
    },
    {
      "type": "model_declared",
      "key": "default",
      "purpose": "language"
    },
    {
      "type": "model_declared",
      "key": "secondary",
      "purpose": "language"
    },
    {
      "type": "context_entered",
      "context": "child"
    },
    {
      "type": "context_exited",
      "context": "child"
    },
    {
      "type": "model_declared",
      "key": "tertiary",
      "purpose": "embedding"
    },
    {
      "type": "exec_exited"
    }
  ],
  "children": [
    {
      "type": "context",
      "scope": {
        "key": "child",
        "events": [
          {
            "type": "model_declared",
            "key": "child_a",
            "purpose": "language"
          },
          {
            "type": "model_declared",
            "key": "child_b",
            "purpose": "embedding"
          }
        ],
        "children": []
      }
    }
  ]
}"
`;
